# PROYECTO INTEGRADOR: STEAM
## Dataset - output_steam_games.json

## 1) Ingesta de datos

Se descomprimen cada uno de los archivos previo a importarlos con pandas.

In [3]:
#Importamos las librerías que utilizaremos a lo largo de nuestro proyecto
import pandas as pd
from pandas import json_normalize
from textblob import TextBlob
import json
import ast 


Importamos la data del archivo output_steam_games.json.

In [4]:
#Importamos en un dataframe la información sobre los juegos de Steam
with open("Data/output_steam_games.json",'r') as f:
    steam_games=[json.loads(line) for line in f]
df_steam_games=pd.DataFrame(steam_games)


In [5]:
#Observamos el dataframe
df_steam_games.head()


publisher genres app_name title  url release_date tags reviews_url  \
0       NaN    NaN      NaN   NaN  NaN          NaN  NaN         NaN   
1       NaN    NaN      NaN   NaN  NaN          NaN  NaN         NaN   
2       NaN    NaN      NaN   NaN  NaN          NaN  NaN         NaN   
3       NaN    NaN      NaN   NaN  NaN          NaN  NaN         NaN   
4       NaN    NaN      NaN   NaN  NaN          NaN  NaN         NaN   

   discount_price specs price early_access   id metascore developer  \
0             NaN   NaN   NaN          NaN  NaN       NaN       NaN   
1             NaN   NaN   NaN          NaN  NaN       NaN       NaN   
2             NaN   NaN   NaN          NaN  NaN       NaN       NaN   
3             NaN   NaN   NaN          NaN  NaN       NaN       NaN   
4             NaN   NaN   NaN          NaN  NaN       NaN       NaN   

             user_id           steam_id  \
0  76561197970982479  76561197970982479   
1            js41637  76561198035864385   
2          evcentric  76561198007712555   
3         Riot-Punch  76561197963445855   
4              doctr  76561198002099482   

                                               items  items_count  
0  [{'item_id': '10', 'item_name': 'Counter-Strik...        277.0  
1  [{'item_id': '10', 'item_name': 'Counter-Strik...        888.0  
2  [{'item_id': '1200', 'item_name': 'Red Orchest...        137.0  
3  [{'item_id': '10', 'item_name': 'Counter-Strik...        328.0  
4  [{'item_id': '300', 'item_name': 'Day of Defea...        541.0

In [6]:
#Observamos las dimensiones de nuestro dataframe
df_steam_games.shape

(120445, 19)

## 2) Tratamiento de los datos

### 2.1) Dataframe: df_steam_games

Podemos observar que en el Dataframe df_steam_games los primeros 88310 registros, corresponden con los 88310 registros que se encuentran en el dataframe df_users_items, por lo tanto podemos proceder a eliminarlos.

In [7]:
#Verificamos los primeros 88310 datos y comparamos con df_items
df_steam_games.iloc[0:10]

publisher genres app_name title  url release_date tags reviews_url  \
0       NaN    NaN      NaN   NaN  NaN          NaN  NaN         NaN   
1       NaN    NaN      NaN   NaN  NaN          NaN  NaN         NaN   
2       NaN    NaN      NaN   NaN  NaN          NaN  NaN         NaN   
3       NaN    NaN      NaN   NaN  NaN          NaN  NaN         NaN   
4       NaN    NaN      NaN   NaN  NaN          NaN  NaN         NaN   
5       NaN    NaN      NaN   NaN  NaN          NaN  NaN         NaN   
6       NaN    NaN      NaN   NaN  NaN          NaN  NaN         NaN   
7       NaN    NaN      NaN   NaN  NaN          NaN  NaN         NaN   
8       NaN    NaN      NaN   NaN  NaN          NaN  NaN         NaN   
9       NaN    NaN      NaN   NaN  NaN          NaN  NaN         NaN   

   discount_price specs price early_access   id metascore developer  \
0             NaN   NaN   NaN          NaN  NaN       NaN       NaN   
1             NaN   NaN   NaN          NaN  NaN       NaN       NaN   
2             NaN   NaN   NaN          NaN  NaN       NaN       NaN   
3             NaN   NaN   NaN          NaN  NaN       NaN       NaN   
4             NaN   NaN   NaN          NaN  NaN       NaN       NaN   
5             NaN   NaN   NaN          NaN  NaN       NaN       NaN   
6             NaN   NaN   NaN          NaN  NaN       NaN       NaN   
7             NaN   NaN   NaN          NaN  NaN       NaN       NaN   
8             NaN   NaN   NaN          NaN  NaN       NaN       NaN   
9             NaN   NaN   NaN          NaN  NaN       NaN       NaN   

                    user_id           steam_id  \
0         76561197970982479  76561197970982479   
1                   js41637  76561198035864385   
2                 evcentric  76561198007712555   
3                Riot-Punch  76561197963445855   
4                     doctr  76561198002099482   
5  MinxIsBetterThanPotatoes  76561198004744620   
6                NitemarePK  76561197990951820   
7                themanwich  76561198035296505   
8                 maplemage  76561198026584251   
9                    Wackky  76561198039117046   

                                               items  items_count  
0  [{'item_id': '10', 'item_name': 'Counter-Strik...        277.0  
1  [{'item_id': '10', 'item_name': 'Counter-Strik...        888.0  
2  [{'item_id': '1200', 'item_name': 'Red Orchest...        137.0  
3  [{'item_id': '10', 'item_name': 'Counter-Strik...        328.0  
4  [{'item_id': '300', 'item_name': 'Day of Defea...        541.0  
5  [{'item_id': '50', 'item_name': 'Half-Life: Op...        371.0  
6  [{'item_id': '240', 'item_name': 'Counter-Stri...        304.0  
7  [{'item_id': '220', 'item_name': 'Half-Life 2'...        258.0  
8  [{'item_id': '240', 'item_name': 'Counter-Stri...        629.0  
9                                                 []          0.0

In [8]:
#Ultimo dato a eliminar del dataframe
df_steam_games.iloc[88309]

publisher                       NaN
genres                          NaN
app_name                        NaN
title                           NaN
url                             NaN
release_date                    NaN
tags                            NaN
reviews_url                     NaN
discount_price                  NaN
specs                           NaN
price                           NaN
early_access                    NaN
id                              NaN
metascore                       NaN
developer                       NaN
user_id           edward_tremethick
steam_id          76561198331598578
items                            []
items_count                     0.0
Name: 88309, dtype: object

In [9]:
#Tomamos el que debería ser el primer dato de nuestro dataframe
df_steam_games.iloc[88310]

publisher                                                 Kotoshiro
genres                [Action, Casual, Indie, Simulation, Strategy]
app_name                                        Lost Summoner Kitty
title                                           Lost Summoner Kitty
url               http://store.steampowered.com/app/761140/Lost_...
release_date                                             2018-01-04
tags                  [Strategy, Action, Indie, Casual, Simulation]
reviews_url       http://steamcommunity.com/app/761140/reviews/?...
discount_price                                                 4.49
specs                                               [Single-player]
price                                                          4.99
early_access                                                  False
id                                                           761140
metascore                                                       NaN
developer                                       

In [10]:
#Elimino los primeros 88310 registros.
df_steam_games = df_steam_games.drop(index=range(88310))

In [11]:
#El primer dato de mi dataframe es el dato 88310 mostrado anteriormente. 
df_steam_games.iloc[0] 

publisher                                                 Kotoshiro
genres                [Action, Casual, Indie, Simulation, Strategy]
app_name                                        Lost Summoner Kitty
title                                           Lost Summoner Kitty
url               http://store.steampowered.com/app/761140/Lost_...
release_date                                             2018-01-04
tags                  [Strategy, Action, Indie, Casual, Simulation]
reviews_url       http://steamcommunity.com/app/761140/reviews/?...
discount_price                                                 4.49
specs                                               [Single-player]
price                                                          4.99
early_access                                                  False
id                                                           761140
metascore                                                       NaN
developer                                       

In [12]:
#Verificamos las nuevas dimensiones de nuestro dataframe
df_steam_games.shape

(32135, 19)

In [13]:
#Reseteamos los indices
df_steam_games=df_steam_games.reset_index(drop=True)

In [14]:
#Sumamos la cantidad de valores nulos que hay en cada una de las columnas del df_steam_games 
df_steam_games.isnull().sum()

publisher          8052
genres             3283
app_name              2
title              2050
url                   0
release_date       2067
tags                163
reviews_url           2
discount_price    31910
specs               670
price              1377
early_access          0
id                    2
metascore         29458
developer          3299
user_id           32135
steam_id          32135
items             32135
items_count       32135
dtype: int64

Vemos que las últimas columnas del dataframe (user_id, steam_id, items, items_count) están vacías, por lo que vamos a proceder a eliminarlas. 

In [15]:
#Eliminamos las columnas sin información
df_steam_games=df_steam_games.drop(columns=['user_id','steam_id', 'items', 'items_count'], axis=1)


In [16]:
#Verificamos las nuevas dimensiones del dataframe
df_steam_games.shape

(32135, 15)

Vamos a eliminar algunas de las columnas del dataframe que no usaremos a lo largo de este proyecto.

In [17]:
#Eliminamos columnas que no utilizaremos para liberar espacio
df_steam_games=df_steam_games.drop(columns=['url','tags','reviews_url','discount_price','metascore','specs','early_access'])

In [18]:
#Verificamos nuevamente las dimensiones del dataframe
df_steam_games.shape

(32135, 8)

In [19]:
#Tenemos solo 2 valores donde app_name no tiene valor
df_steam_games[df_steam_games['app_name'].isna()]

publisher           genres app_name title release_date  price      id  \
74         NaN              NaN      NaN   NaN          NaN  19.99     NaN   
2580       NaN  [Action, Indie]      NaN   NaN   2014-08-26    NaN  317160   

     developer  
74         NaN  
2580       NaN

En lo que refiere al registro 74, podemos observar que no contamos con los datos más importantes como app_name, id, specs entre otras, intentamos a través del URL identificar el valor, pero como el URL nos dirige a un landpage y no nos permite identificar la aplicación, procedemos a eliminar el registro 74. 

In [20]:
df_steam_games=df_steam_games.drop(74)

En lo que refiere al registro 2580, tenemos algo más de información, como el URL, la fecha de release, id, entre otras. Vamos a proceder a buscar a partir de la URL (http://store.steampowered.com/app/317160/_/) el nombre de la app y el título, y lo vamos a actualizar en el dataframe. 

In [21]:
#Actualizamos tanto el app_name como el title del registro 2580
df_steam_games.loc[2580, 'app_name'] = "Duet" 
df_steam_games.loc[2580, 'title'] = "Duet"

In [22]:
df_steam_games.loc[2580]

publisher                   NaN
genres          [Action, Indie]
app_name                   Duet
title                      Duet
release_date         2014-08-26
price                       NaN
id                       317160
developer                   NaN
Name: 2580, dtype: object

Además del app_name vemos que hay 2 registros faltantes en la columna id, uno era el registro 74 eliminado, el otro es el registro 30961.

In [23]:
df_steam_games[df_steam_games['id'].isna()]

publisher               genres  \
30961  Warner Bros. Interactive Entertainment, Feral ...  [Action, Adventure]   

                                             app_name  \
30961  Batman: Arkham City - Game of the Year Edition   

                                                title release_date  price  \
30961  Batman: Arkham City - Game of the Year Edition   2012-09-07  19.99   

        id                                   developer  
30961  NaN  Rocksteady Studios,Feral Interactive (Mac)

A partir del URL del registro 30961 (http://store.steampowered.com/app/200260) podemos determinar que el id es  200260, por lo que vamos a proceder a actualizarlo en nuestro dataframe. 

In [24]:
#Actualizamos el valor del id
df_steam_games.loc[30961, 'id'] = 200260

In [25]:
df_steam_games.loc[30961]

publisher       Warner Bros. Interactive Entertainment, Feral ...
genres                                        [Action, Adventure]
app_name           Batman: Arkham City - Game of the Year Edition
title              Batman: Arkham City - Game of the Year Edition
release_date                                           2012-09-07
price                                                       19.99
id                                                         200260
developer              Rocksteady Studios,Feral Interactive (Mac)
Name: 30961, dtype: object

In [26]:
#Volvemos a verificar los valores nulo
df_steam_games.isnull().sum()

publisher       8051
genres          3282
app_name           0
title           2048
release_date    2066
price           1377
id                 0
developer       3298
dtype: int64

Podemos comprobar que tanto el id, como el release_date figuran como objeto aunque deberían ser un entero y una fecha respectivamente, procederemos a realizar dicha modificación.


In [27]:
#Modificamos el tipo de dato de id
df_steam_games["id"] = df_steam_games["id"].astype(int)

In [28]:
#Verificamos que se haya modificado el tipo de dato de la columna id
df_steam_games["id"].info()

<class 'pandas.core.series.Series'>
Index: 32134 entries, 0 to 32134
Series name: id
Non-Null Count  Dtype
--------------  -----
32134 non-null  int32
dtypes: int32(1)
memory usage: 1.4 MB


In [29]:
#Modificamos el tipo de dato de release_date
df_steam_games['release_date'] = pd.to_datetime(df_steam_games['release_date'], errors='coerce')

In [30]:
#Verificamos que se haya modificado el tipo de dato de la columna release_date
df_steam_games["release_date"].info()

<class 'pandas.core.series.Series'>
Index: 32134 entries, 0 to 32134
Series name: release_date
Non-Null Count  Dtype         
--------------  -----         
29783 non-null  datetime64[ns]
dtypes: datetime64[ns](1)
memory usage: 1.5 MB


Según las funciones solicitadas, solamente vamos a necesitar del release_date, el año. Por lo tanto, crearemos una columna donde solo se incluya el año a partir de la información de la columna release_date.

In [31]:
#Creamos la columna release_year con el año de la columna release_date
df_steam_games["release_year"] = df_steam_games["release_date"].dt.year

In [32]:
#Controlamos los valores unicos que aparecen en la columna release_year
df_steam_games["release_year"].unique()

array([2018., 2017.,   nan, 1997., 1998., 2016., 2006., 2005., 2003.,
       2007., 2002., 2000., 1995., 1996., 1994., 2001., 1993., 2004.,
       1999., 2008., 2009., 1992., 1989., 2010., 2011., 2013., 2012.,
       2014., 1983., 1984., 2015., 1990., 1988., 1991., 1985., 1982.,
       1987., 1981., 1986., 2021., 2019., 1975., 1970., 1980.])

In [33]:
df_steam_games.shape

(32134, 9)

Luego del análisis de los datos, como varias de las funciones me solicitan que use el año  de lanzamiento, procedemos a eliminar los registros donde release_year es nulo, mantener los que no son nulos. 

In [34]:
#Mantenemos los valores donde el release_year sea distinto de nan.
df_steam_games = df_steam_games[df_steam_games['release_year'].notna()]

In [35]:
#Verificamos las dimensiones del dataframe
df_steam_games.shape

(29783, 9)

In [36]:
#Cambiamos el tipo de dato de la columna a entero
df_steam_games['release_year'] = df_steam_games["release_year"].astype(int)

In [37]:
#Verificamos el cambio de tipo de dato para la columna release_year
df_steam_games['release_year'].info()

<class 'pandas.core.series.Series'>
Index: 29783 entries, 0 to 32133
Series name: release_year
Non-Null Count  Dtype
--------------  -----
29783 non-null  int32
dtypes: int32(1)
memory usage: 349.0 KB


Vamos ahora a trabajar con la columna genres.

In [38]:
#Verificamos la suma de los valores nulos de la columna genres
df_steam_games['genres'].isnull().sum()

1234

En total tenemos 1234 registros que son nulos en la columna genres. Decidimos eliminarlos basándonos en que varias de las funciones nos piden un género no nulo para hacer las consultas. 

In [39]:
#Mantenemos los valores de genres que no son nulos
df_steam_games = df_steam_games[df_steam_games['genres'].notna()]

In [40]:
#Verifico las dimensiones del dataframe
df_steam_games.shape

(28549, 9)

Además, en la columna genres podemos evidenciar que tenemos una lista de elementos, por lo tanto vamos a proceder a usar el comando explode para que cada elemento de la lista se constituya en un único registro.

In [41]:
#Observamos como se compone la columna genres
df_steam_games['genres'].head()

0        [Action, Casual, Indie, Simulation, Strategy]
1                 [Free to Play, Indie, RPG, Strategy]
2    [Casual, Free to Play, Indie, Simulation, Sports]
3                          [Action, Adventure, Casual]
5                      [Action, Adventure, Simulation]
Name: genres, dtype: object

In [42]:
#Usamos el comando explode para poder separar cada elemento de la lista en un registro
df_games = df_steam_games.explode('genres')

In [43]:
#Revisamos como ha quedado nuestro dataframe
df_games.head()

publisher      genres             app_name                title  \
0  Kotoshiro      Action  Lost Summoner Kitty  Lost Summoner Kitty   
0  Kotoshiro      Casual  Lost Summoner Kitty  Lost Summoner Kitty   
0  Kotoshiro       Indie  Lost Summoner Kitty  Lost Summoner Kitty   
0  Kotoshiro  Simulation  Lost Summoner Kitty  Lost Summoner Kitty   
0  Kotoshiro    Strategy  Lost Summoner Kitty  Lost Summoner Kitty   

  release_date price      id  developer  release_year  
0   2018-01-04  4.99  761140  Kotoshiro          2018  
0   2018-01-04  4.99  761140  Kotoshiro          2018  
0   2018-01-04  4.99  761140  Kotoshiro          2018  
0   2018-01-04  4.99  761140  Kotoshiro          2018  
0   2018-01-04  4.99  761140  Kotoshiro          2018

In [44]:
print('Las dimensiones del dataframe df_steam_games son:', df_steam_games.shape)
print('Las dimensiones del dataframe df_games son:', df_games.shape)

Las dimensiones del dataframe df_steam_games son: (28549, 9)
Las dimensiones del dataframe df_games son: (70704, 9)


In [45]:
#Evidenciamos los géneros (y los vacíos)
df_games['genres'].unique()

array(['Action', 'Casual', 'Indie', 'Simulation', 'Strategy',
       'Free to Play', 'RPG', 'Sports', 'Adventure', 'Racing',
       'Massively Multiplayer', 'Early Access',
       'Animation &amp; Modeling', 'Video Production', 'Web Publishing',
       'Education', 'Software Training', 'Utilities',
       'Design &amp; Illustration', 'Audio Production', 'Photo Editing',
       'Accounting'], dtype=object)

In [46]:
df_games.reset_index(drop=True, inplace=True)

Finalmente decidimos eliminar otras columnas como: publisher, app_name, release_date, price y developer

In [47]:
#Elimino las columnas
df_games=df_games.drop(columns=['publisher','app_name', 'release_date','price','developer'])

In [48]:
#Verificamos el dataframe antes de exportarlo
df_games.head()

genres                title      id  release_year
0      Action  Lost Summoner Kitty  761140          2018
1      Casual  Lost Summoner Kitty  761140          2018
2       Indie  Lost Summoner Kitty  761140          2018
3  Simulation  Lost Summoner Kitty  761140          2018
4    Strategy  Lost Summoner Kitty  761140          2018

In [49]:
#Exportamos a csv y lo guardamos en la carpeta Data
df_games.to_csv('Data/games.csv', index=False)